### DISCUSSION TITLE
The example xmls provided are giving errors in schematron validation

### DISCUSSION THREAD

**Discussion by rgannu** [OP-TED/eForms-SDK#75] [Q&A] [answered]
I have taken the examples and schematrons from the 0.7.0 release of eForms-SDK. 

I use the [ph-schematron](https://github.com/phax/ph-schematron) - Java library to validate XML documents according to Schematron (ph-schematron-pure)

When using the schematrons (`entry.sch`) to validate the examples I get errors in the following:
- cn_24_maximal.xml
```
[SVRLFailedAssert@0x10db6131: DiagnosticRefs=[[DiagnosticReference@0x6c6017b9: content=[[[Text@0x4730e0f0: content=[cbc:ProcessReason]]; fpi=null; icon=null; lang=null; see=null; space=null]]; diagnostic=BT-1351-Procedure; fpi=null; icon=null; lang=null; see=null; space=null]]; ID=BR-BT-01351-0054; Text=rule|message|BR-BT-01351-0054; Location=/ContractNotice/cac:TenderingProcess/cac:ProcessJustification; Test=count(cbc:ProcessReason) = 0 or (cbc:ProcessReasonCode[@listName='accelerated-procedure']/normalize-space(text()) = 'accelerated-procedure-justification'); Role=ERROR; Flag=ERROR]
```

- cn_24_minimal.xml
````
[SVRLFailedAssert@0x47406941: DiagnosticRefs=[[DiagnosticReference@0x61ecbee9: content=[[[Text@0x1bb740f2: content=[cac:MediationParty/cac:PartyIdentification/cbc:ID]]; fpi=null; icon=null; lang=null; see=null; space=null]]; diagnostic=OPT-301-Lot-Mediator; fpi=null; icon=null; lang=null; see=null; space=null]]; ID=BR-OPT-00301-1180; Text=rule|message|BR-OPT-00301-1180; Location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AppealTerms; Test=count(cac:MediationParty/cac:PartyIdentification/cbc:ID) > 0; Role=WARN; Flag=WARN]
[SVRLFailedAssert@0x2538bc06: DiagnosticRefs=[[DiagnosticReference@0x61f6d381: content=[[[Text@0x656922a0: content=[cbc:WebsiteURI]]; fpi=null; icon=null; lang=null; see=null; space=null]]; diagnostic=BT-505-Organization-Company; fpi=null; icon=null; lang=null; see=null; space=null]]; ID=BR-BT-00505-0022; Text=rule|message|BR-BT-00505-0022; Location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[0]/efac:Company; Test=count(cbc:WebsiteURI) > 0 or not(cac:PartyName/cbc:Name); Role=WARN; Flag=WARN]
[SVRLFailedAssert@0x44784e2f: DiagnosticRefs=[[DiagnosticReference@0x2440022a: content=[[[Text@0x737db7f8: content=[cac:Contact/cbc:Telefax]]; fpi=null; icon=null; lang=null; see=null; space=null]]; diagnostic=BT-739-Organization-Company; fpi=null; icon=null; lang=null; see=null; space=null]]; ID=BR-BT-00739-0022; Text=rule|message|BR-BT-00739-0022; Location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[0]/efac:Company; Test=count(cac:Contact/cbc:Telefax) > 0 or not(cac:PartyName/cbc:Name); Role=WARN; Flag=WARN]
[SVRLFailedAssert@0x5f2de715: DiagnosticRefs=[[DiagnosticReference@0x5922d3e9: content=[[[Text@0x7d57dbb5: content=[cbc:WebsiteURI]]; fpi=null; icon=null; lang=null; see=null; space=null]]; diagnostic=BT-505-Organization-Company; fpi=null; icon=null; lang=null; see=null; space=null]]; ID=BR-BT-00505-0022; Text=rule|message|BR-BT-00505-0022; Location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[1]/efac:Company; Test=count(cbc:WebsiteURI) > 0 or not(cac:PartyName/cbc:Name); Role=WARN; Flag=WARN]
[SVRLFailedAssert@0x630b6190: DiagnosticRefs=[[DiagnosticReference@0x532e27ab: content=[[[Text@0x1cdc1bbc: content=[cac:Contact/cbc:Telefax]]; fpi=null; icon=null; lang=null; see=null; space=null]]; diagnostic=BT-739-Organization-Company; fpi=null; icon=null; lang=null; see=null; space=null]]; ID=BR-BT-00739-0022; Text=rule|message|BR-BT-00739-0022; Location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[1]/efac:Company; Test=count(cac:Contact/cbc:Telefax) > 0 or not(cac:PartyName/cbc:Name); Role=WARN; Flag=WARN]
```

Others which fail in schematron validation are
- cn_24_maximal_100_lots.xml
- cn_24_multilingual.xml
- pin-buyer_24_minimal.xml
- cn_25.xml
- cn_24_open_accel.xml
- t02_ESP.xml
- pin-cfc_25.xml
- can_24_minimal.xml
- can_24_maximal.xml
- CEI.xml

Any reason why the schematron rules aren't validating these xmls?

**Comment by bertrand-lorentz** [comment, marked as answer]
Thanks for your feedback.

There are 2 different issues at play for the various failed asserts you see.

**1/ role="WARN"**
There are several examples that cause assert failures, but with `role="WARN"`.
The rules that have `role="WARN"` do not make the notice invalid, they are here to point to a possible or potential issue.
You can find those warnings in the validation reports corresponding to the examples, for instance, for cn_24_minimal.xml:
https://github.com/OP-TED/eForms-SDK/blob/develop/examples/reports/cn_24_minimal.svrl#L108

**2/ Incorrect rules**
The failure you indicate in `cn_24_maximal.xml` seems to be caused by a set of incorrect rules for field BT-1351-Procedure, that affect several other notice subtypes.
We will correct those rules in the next version of the SDK.

We didn't notice those rules were incorrect, as those errors do not appear in the corresponding validation reports in the SDK. That's because ph-schematron-pure has a different behavior than the Schematron implementation we are using to generated those reports: the incorrect rule happens to not be execute by our implementation, due to the behavior described here:
https://schematron.com/document/144.html

As we do want all rules to be executed, we will have to modify our Schematron so that they are indeed executed with all Schematron implementations.

I've executed the example XML files with ph-schematron, and it seems that point 1/ and 2/ cover the failures in the examples you listed.

**Comment by rgannu** [comment]
> As we do want all rules to be executed, we will have to modify our Schematron so that they are indeed executed with all Schematron implementations.

Thanks for the update. Can you please indicate in which release can we expect the schmatron files to be updated?

**Reply by rousso** [reply]
Hi @rgannu,

We have not yet set a date for the next release of the SDK, However here is what I can already tell you about it:
- I expect that SDK 1.1.0 will be released by September 9.
- After that, we intend to release updates to the eForms metadata as often as possible. 
- We do not plan to change the major version of the SDK in the foreseeable future, so all our coming releases will be backwards compatible with 1.0.0 and will only update/fix/improve the eForms metadata contained in the SDK.

As we are currently trying to streamline the process for future releases, we will update our [roadmap](https://docs.ted.europa.eu/home/eforms/roadmap/index.html) as soon as we have more specific information to share. 

For the semantics of SDK versioning and how to automatically "consume" new backwards compatible releases please also see our [versioning information](https://docs.ted.europa.eu/eforms/1.0.0/versioning.html) page.

**Comment by DKamerbeek** [comment]
My project is also using [ph-schematron](https://github.com/phax/ph-schematron) lib. I can confirm that SDK 1.0.0 also has invalid xml examples that should be valid. I filtered out the WARN assertions, these are the results (ran with static and dynamic schema's):

file: **change-cn_24_suppliers.xml**
errors:
- location=/ContractNotice; error=rule|text|BR-BT-00701-0022
- location=/ContractNotice/cac:ProcurementProject/cac:RealizedLocation/cac:Address; error=rule|text|BR-BT-05071-0206
- location=/ContractNotice/cac:ProcurementProject/cac:RealizedLocation/cac:Address; error=rule|text|BR-BT-05141-0022
- location=/ContractNotice/cac:ProcurementProjectLot/cac:ProcurementProject/cac:RealizedLocation; error=rule|text|BR-BT-05071-0168
- location=/ContractNotice/cac:ProcurementProjectLot/cac:ProcurementProject/cac:RealizedLocation; error=rule|text|BR-BT-05141-0124
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingProcess; error=rule|text|BR-BT-00131-0022
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms; error=rule|text|BR-BT-00018-0022
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms; error=rule|text|BR-BT-00098-0022
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AwardingTerms/cac:AwardingCriterion/cac:SubordinateAwardingCriterion[0]; error=rule|text|BR-BT-00540-0074
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AwardingTerms/cac:AwardingCriterion/cac:SubordinateAwardingCriterion[0]; error=rule|text|BR-BT-00734-0074
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AwardingTerms/cac:AwardingCriterion/cac:SubordinateAwardingCriterion[1]; error=rule|text|BR-BT-00540-0074
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AwardingTerms/cac:AwardingCriterion/cac:SubordinateAwardingCriterion[1]; error=rule|text|BR-BT-00734-0074
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AwardingTerms/cac:AwardingCriterion/cac:SubordinateAwardingCriterion[0]/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension; error=rule|text|BR-BT-00541-0073
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:AwardingTerms/cac:AwardingCriterion/cac:SubordinateAwardingCriterion[1]/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension; error=rule|text|BR-BT-00541-0073
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms/cac:CallForTendersDocumentReference; error=rule|text|BR-BT-00015-0073
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Changes/efac:Change[0]; error=rule|text|BR-BT-13716-0071
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Changes/efac:Change[1]; error=rule|text|BR-BT-13716-0071
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[1]/efac:Company; error=rule|text|BR-BT-00507-0022
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[1]/efac:Company; error=rule|text|BR-BT-00512-0022
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[2]/efac:Company; error=rule|text|BR-BT-00507-0022
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[2]/efac:Company; error=rule|text|BR-BT-00512-0022
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[3]/efac:Company; error=rule|text|BR-BT-00507-0022
- location=/ContractNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[3]/efac:Company; error=rule|text|BR-BT-00512-0022
- location=/ContractNotice; error=rule|text|BR-BT-00002-0104
- location=/ContractNotice/cac:ProcurementProjectLot; error=rule|text|BR-BT-00137-0200
- location=/ContractNotice/cac:ProcurementProjectLot/cac:TenderingTerms; error=rule|text|BR-BT-00743-0100
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
file: **can_24_maximal.xml**
errors:
- location=/ContractAwardNotice/cac:TenderingProcess/cac:ProcessJustification; error=rule|text|BR-BT-01351-0036
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
file: **cn_24_nego_accel.xml**
errors:
- location=/ContractNotice/cac:TenderingProcess/cac:ProcessJustification; error=rule|text|BR-BT-01351-0054
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
file: **t02_ESP.xml**
errors:
- location=/ContractAwardNotice/ext:UBLExtensions/ext:UBLExtension/ext:ExtensionContent/efext:EformsExtension/efac:Organizations/efac:Organization[2]/efac:Company; error=rule|text|BR-BT-00165-0041
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
file: **cn_24_maximal.xml**
errors:
- location=/ContractNotice/cac:TenderingProcess/cac:ProcessJustification; error=rule|text|BR-BT-01351-0054
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
file: **cn_24_open_accel.xml**
errors:
- location=/ContractNotice/cac:TenderingProcess/cac:ProcessJustification; error=rule|text|BR-BT-01351-0054
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
file: **cn_24_maximal_100_lots.xml**
errors:
- location=/ContractNotice/cac:TenderingProcess/cac:ProcessJustification; error=rule|text|BR-BT-01351-0054
